flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/sources/ProjectableTableSource.scala

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.sources
+
+// TODO FLINK-3848
+/**
+  * Defines a TableSource that can return a projected version of the table.
+  *
+  * @tparam T The return type of the [[TableSource]].
+  */
+trait ProjectableTableSource[T] extends TableSource[T] {
+
+  /** Returns a [[TableSource]] with ability to project fields */
+  def setProjection(fields: Array[Int]): Unit
```

Comment:
I would like the definition of `ProjectableTableSource `  be like this : https://github.com/fhueske/flink/blob/tableProjectSource/flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/sources/ProjectableTableSource.scala

The `setProjection` should return a copy TableSource. So the action of project pushdown won't affect other queries using this TableSource. Considering this case: 

```scala
  tEnv.registerTableSource("csvTable", csvTable)
    val right = tEnv
      .ingest("csvTable")
      .select('id as 'rid, 'score)

    val table = tEnv
      .ingest("csvTable")
      .select('id, 'last)
      .join(right, 'id == 'rid)
```